### PR TITLE
core(complex): constexpr

### LINKED ISSUE #6281: `Kokkos::complex` should have `constexpr` constructor as well
If I'm not mistaken, according to https://en.cppreference.com/w/cpp/numeric/complex/complex there are several constructors of `std::complex` that are declared `constexpr` already in the `c++14` standard.

Looking at
https://github.com/kokkos/kokkos/blob/7e6871f6a163d3ffff362ff4c64c3d4616d55f62/core/src/Kokkos_Complex.hpp#L109
it seems it is not the case for `Kokkos::complex`, and `constexpr` needs to be explicitly added.

Here is a small code to test `constexpr`-ness:
```c++
template<class Lambda, int=(Lambda{}(), 0)>
constexpr bool is_constexpr(Lambda) { return true; }
constexpr bool is_constexpr(...) { return false; }

if constexpr(is_constexpr([]{ std::complex{1.}; }))
        std::cout << "std::complex is constexpr" << std::endl;
    else
        std::cout << "unexpected !" << std::endl;

if constexpr(is_constexpr([]{ Kokkos::complex{1.}; }))
        std::cout << "Kokkos::complex is constexpr" << std::endl;
    else
        std::cout << "unexpected !" << std::endl;
```

This should show (at least for `g++ 11.3.0`):
```text
std::complex is constexpr
unexpected !
```

If you agree on the changes, I can open a PR.

### PR DESCRIPTION
**Pull request by romintomasetti**
This fixes #6281.

### CODE REVIEW

**Review by masterleinad — APPROVED** [review]

File: core/src/Kokkos_Complex.hpp [hunk 1]
```diff
@@ -64,10 +64,10 @@ class
 
   //! Copy constructor.
   KOKKOS_DEFAULTED_FUNCTION
-  complex(const complex&) noexcept = default;
+  constexpr complex(const complex&) noexcept = default;
```
Comment (JBludau) [inline core/src/Kokkos_Complex.hpp:67]: do we even need to make it `constexpr`? thinking of https://stackoverflow.com/questions/20810378/should-i-mark-a-compiler-generated-constructor-as-constexpr

Comment (romintomasetti) [reply core/src/Kokkos_Complex.hpp:67]: Indeed, I went a bit fast on those. Good catch!

**Comment by masterleinad** [comment]
```
In file included from /builds/kokkos/kokkos/build/core/unit_test/hip/TestHIP_Complex.cpp:2:
/builds/kokkos/kokkos/core/unit_test/TestComplex.hpp:370:29: error: unused variable 'b' [-Werror,-Wunused-variable]
  370 |     Kokkos::complex<double> b(3.25, 5.75);
      |                             ^
```

**Review by crtrott — APPROVED** [review]

File: core/unit_test/TestComplex.hpp [hunk 2]
```diff
@@ -688,6 +687,24 @@ TEST(TEST_CATEGORY, complex_structured_bindings) {
...
+  [[maybe_unused]] Kokkos::complex<double> from_both{1.2, 3.4};
+  [[maybe_unused]] Kokkos::complex<double> from_none{};
+
+  [[maybe_unused]] auto from_copy_assign = from_both;
+  [[maybe_unused]] auto from_copy_constr(from_both);
+
+  [[maybe_unused]] auto from_move_assign = std::move(from_both);
+  [[maybe_unused]] auto from_move_constr(std::move(from_single));
+
+  [[maybe_unused]] Kokkos::complex<double> from_real = 4.;
```
Comment (dalg24) [inline core/unit_test/TestComplex.hpp:701]: I would want to see some static assertions about the real and imaginary parts in that test

File: core/unit_test/TestComplex.hpp [hunk 3]
```diff
@@ -688,6 +687,24 @@ TEST(TEST_CATEGORY, complex_structured_bindings) {
   test.testit();
 }
 
+constexpr bool basics_are_constexpr() {
```
Comment (dalg24) [inline core/unit_test/TestComplex.hpp:690]: What does "basics" mean?

**Review by dalg24 — COMMENTED** [review]
I don't really see the point if none of the other functionalities can appear in constant expressions.
Also I expect @nliber was working on this

File: core/unit_test/TestComplex.hpp (hunk 2 above)
Comment (romintomasetti) [reply core/unit_test/TestComplex.hpp]: Done, the test is now rock solid.

File: core/unit_test/TestComplex.hpp (hunk 3 above)
Comment (romintomasetti) [reply core/unit_test/TestComplex.hpp]: I've renamed it `construction_is_constexpr`, OK with you ?

**Comment by romintomasetti** [comment]
@dalg24 All green :green_circle:

File: core/unit_test/TestComplex.hpp [hunk 4]
```diff
@@ -688,6 +687,40 @@ TEST(TEST_CATEGORY, complex_structured_bindings) {
...
+constexpr bool construction_is_constexpr() {
+  [[maybe_unused]] constexpr Kokkos::complex<double> from_single{1.2};
+  [[maybe_unused]] constexpr Kokkos::complex<double> from_both{1.2, 3.4};
+  [[maybe_unused]] constexpr Kokkos::complex<double> from_none{};
+
+  CHECK_COMPLEX(from_single, 1.2, 0.);
+  CHECK_COMPLEX(from_both, 1.2, 3.4);
+  CHECK_COMPLEX(from_none, 0., 0.);
+
+  [[maybe_unused]] constexpr auto from_copy_assign = from_both;
```
Comment (dalg24) [inline core/unit_test/TestComplex.hpp:703]: This is not a copy assign, this is going through the copy constructor.

File: core/unit_test/TestComplex.hpp [hunk 5]
```diff
@@ -688,6 +687,40 @@ TEST(TEST_CATEGORY, complex_structured_bindings) {
...
+  [[maybe_unused]] constexpr auto from_copy_constr(from_both);
+
+  CHECK_COMPLEX(from_copy_assign, 1.2, 3.4);
+  CHECK_COMPLEX(from_copy_constr, 1.2, 3.4);
+
+  [[maybe_unused]] constexpr auto from_move_assign = std::move(from_both);
```
Comment (dalg24) [inline core/unit_test/TestComplex.hpp:709]: Same here, this uses move constructor, not move assignment operator

File: core/unit_test/TestComplex.hpp [hunk 6]
```diff
@@ -688,6 +687,40 @@ TEST(TEST_CATEGORY, complex_structured_bindings) {
...
+  [[maybe_unused]] constexpr auto from_move_constr(std::move(from_copy_assign));
+
+  CHECK_COMPLEX(from_move_assign, 1.2, 3.4);
+  CHECK_COMPLEX(from_move_constr, 1.2, 3.4);
+
+  [[maybe_unused]] constexpr Kokkos::complex<double> from_real = 4.;
+
+  CHECK_COMPLEX(from_real, 4., 0.);
```
Comment (dalg24) [inline core/unit_test/TestComplex.hpp:717]: Already covered by `from_single` above

File: core/unit_test/TestComplex.hpp (hunk 5 above)
Comment (romintomasetti) [reply core/unit_test/TestComplex.hpp:709]: Indeed, let me correct this.

File: core/unit_test/TestComplex.hpp (hunk 4 above)
Comment (romintomasetti) [reply core/unit_test/TestComplex.hpp:703]: Indeed, let me correct this.

File: core/unit_test/TestComplex.hpp (hunk 6 above)
Comment (romintomasetti) [reply core/unit_test/TestComplex.hpp:717]: Changed it to an assignment indeed.

File: core/unit_test/TestComplex.hpp [hunk 7]
```diff
@@ -688,6 +687,43 @@ TEST(TEST_CATEGORY, complex_structured_bindings) {
   test.testit();
 }
 
+#define CHECK_COMPLEX(_value_, _real_, _imag_) \
+  if (_value_.real() != _real_) return false;  \
+  if (_value_.imag() != _imag_) return false;
+
+constexpr bool construction_is_constexpr() {
```
Comment (dalg24) [inline core/unit_test/TestComplex.hpp:694]:
```suggestion
constexpr bool can_appear_in_constant_expressions() {
```

File: core/unit_test/TestComplex.hpp [hunk 8]
```diff
@@ -688,6 +687,43 @@ TEST(TEST_CATEGORY, complex_structured_bindings) {
...
+  [[maybe_unused]] const Kokkos::complex<double> from_single{1.2};
+  [[maybe_unused]] const Kokkos::complex<double> from_both{1.2, 3.4};
+  [[maybe_unused]] const Kokkos::complex<double> from_none{};
```
Comment (dalg24) [inline core/unit_test/TestComplex.hpp:697]: I would have dropped all the maybe_unused attributes and just cast to void in the CHECK_COMPLEX macro.

File: core/unit_test/TestComplex.hpp [hunk 9]
```diff
@@ -688,6 +687,43 @@ TEST(TEST_CATEGORY, complex_structured_bindings) {
...
+
+  [[maybe_unused]] Kokkos::complex<double> from_real;
+  from_real = 4.;
+
+  CHECK_COMPLEX(from_real, 4., 0.);
+
+  return true;
+}
+
+static_assert(construction_is_constexpr());
```
Comment (dalg24) [inline core/unit_test/TestComplex.hpp:725]: `#undef CHECK_COMPLEX`

**Review by dalg24 — COMMENTED** [review]
I only have minor comments.
Changes are optional

File: core/unit_test/TestComplex.hpp (hunk 7 above)
Comment (romintomasetti) [reply core/unit_test/TestComplex.hpp:694]: :heavy_check_mark:

File: core/unit_test/TestComplex.hpp (hunk 9 above)
Comment (romintomasetti) [reply core/unit_test/TestComplex.hpp:725]: :heavy_check_mark:

**Review by JBludau — APPROVED** [review]
lgtm

**Comment by dalg24** [comment]
Retest this please

**Pull request merged**